### PR TITLE
[Bug Fix] Class Trainers dont steal your money!

### PR DESCRIPTION
**Pull request by fryguy503**
Class trainers log an error when you have reached your max train in the specific skill. They will continue to "Appear" to eat money and practices however they will be returned on logout/in or zone.

### CODE REVIEW

File: zone/client_process.cpp [hunk 1]
```diff
@@ -1691,11 +1696,12 @@ void Client::OPGMTrainSkill(const EQApplicationPacket *app)
 
 		uint16 skilllevel = GetRawSkill(skill);
 
-		if(skilllevel == 0) {
+		if(skilllevel == 0 || (skilllevel == 254 && SkillTrainLevel(skill, GetClass()) <= GetLevel())) {
```
Comment (Kinglykrab) [inline zone/client_process.cpp:1699]: What does checking for `254` do here exactly?

**Comment by Kinglykrab** [comment]
What do the `SetSkill` calls do exactly?

File: zone/client_process.cpp (hunk 1 above)
Comment (fryguy503) [reply zone/client_process.cpp:1699]: Thank you, this was a carry over from the older TAKP code our server used. I do not believe this portion is needed anymore.

**Comment by fryguy503** [comment]
SetSkill is setting the current skill to the value before the attempted increase. SetSkill is sending the packet to the client which dumbly continues to increase the values beyond the max. This makes sure the client is in sync.

File: zone/client_process.cpp (hunk 1 above)
Comment (Kinglykrab) [reply zone/client_process.cpp:1699]: Still need to remove this if we don't need this.

File: zone/client_process.cpp [hunk 2]
```diff
@@ -1552,19 +1552,24 @@ void Client::OPGMTraining(const EQApplicationPacket *app)
...
-		if (GetClass() != trains_class)
+		if (GetClass() != trains_class) {
+			safe_delete(outapp);
 			return;
+		}
 	}
 
 	//you have to be somewhat close to a trainer to be properly using them
-	if(DistanceSquared(m_Position,pTrainer->GetPosition()) > USE_NPC_RANGE2)
+	if(DistanceSquared(m_Position,pTrainer->GetPosition()) > USE_NPC_RANGE2) {
```
Comment (Kinglykrab) [inline zone/client_process.cpp:1569]: `if (`

File: zone/client_process.cpp [hunk 3]
```diff
@@ -1552,19 +1552,24 @@ void Client::OPGMTraining(const EQApplicationPacket *app)
 
 	Mob* pTrainer = entity_list.GetMob(gmtrain->npcid);
 
-	if(!pTrainer || !pTrainer->IsNPC() || pTrainer->GetClass() < Class::WarriorGM || pTrainer->GetClass() > Class::BerserkerGM)
+	if(!pTrainer || !pTrainer->IsNPC() || pTrainer->GetClass() < Class::WarriorGM || pTrainer->GetClass() > Class::BerserkerGM) {
```
Comment (Kinglykrab) [inline zone/client_process.cpp:1555]: `if (`

File: zone/client_process.cpp (hunk 1 above)
Comment (Kinglykrab) [inline zone/client_process.cpp:1699]: `if (`

**Review by Akkadius — APPROVED** [review]

**Pull request merged**